tests/grid/global_ids_03.cc

```diff
@@ -52,7 +52,8 @@ test(int n_refinements, MPI_Comm comm)
           deallog << cell->id() << " -> " << cell->level_subdomain_id() << " "
                   << cell->global_level_cell_index() << std::endl;
 
-      const auto &part = tria.global_level_cell_index_partitioner(l);
+      const Utilities::MPI::Partitioner &part =
+        *tria.global_level_cell_index_partitioner(l).lock();
```

Comment:
I'm late to the party, but I don't really understand the concept that is used here. Maybe it's only my partial understanding of what `weak_ptr.lock()` really does, so I'd like someone to help me with my concern:

In short, I can't make sense of the value `weak_ptr` brings to this context. This change of return type surely helps the user who writes `const auto partitioner = tria.global_active_cell_index_partitioner()`, but that's about it. All other situations I image, most importantly the construction and reinitialization of `LinearAlgebra::distributed::Vector` through an `std::shared_ptr<const Partitioner>`, only get more tedious to write by saying `...._index_partitioner().lock()`. But what do we gain then from the weak pointer? We convert it to a shared pointer immediately, which is never a problem because at the place the triangulation must is still be alive because to allow access to its member. But then inside the vector the shared pointer keeps the partitioner alive, irrespective of the life time of the triangulation.

In my opinion, the weak pointer would make more sense as a member variable somewhere, not as a return value. Apart from `auto` capture of course and possibly educating the user to keep an eye on the lifetime of some objects. However, I think it is clear that we don't want to change the vector interface as a linear algebra object whose lifetime does not depend on where the numbering came from, so where is the value?

Not that I'm completely opposed to returning a weak pointer instead of a shared pointer, but at this specific place it seems to mostly provide theoretical value at the cost of some nuisance in practice. (Not all users will know how to handle a weak pointer and get the desired shared pointer other interfaces expect out of it, so we need to write additional documentation somewhere to educate about the not so well-known features of STL, and so on.)

Reply:
A `std::weak_ptr` would be useful if we would make also the internal partitioner of `LA:p:V`to a `weak_ptr`; once it is used, e.g. for communication, (it is converted to a `shared_ptr` and) it is checked if the partitioner and the triangulation class are still living. 

But as long as we are not using it this way (maybe we never what to use it this way), we won't have any benefit, as far as see it. In our use case, we immediately convert the weak pointer to a shared pointer:
```cpp
LinearAlgebra::parallel::Vector<double> vector(tria.global_active_cell_index_partitioner().lock());
``` 
so that we actually need to write more code (`.lock()`).

I would prefer to use `std::shared_ptr` because of the reason listed in https://github.com/dealii/dealii/pull/11859#issuecomment-792071475. But I see @bangerth's reasoning: this might be not beneficial right now, but maybe in the future!?